pkg/operator/configobservation/cloud/observe_cloud_volume_plugin.go

```diff
@@ -0,0 +1,110 @@
+	"github.com/openshift/library-go/pkg/operator/events"
+)
+
+// ObserveCloudVOlumePlugin fills in the extendedArguments.external-cloud-volume-plugin with the value of the current
+// platform type, only when the cluster is running an external cloud provider and there is a supported in-tree volume
+// plugin.
+func ObserveCloudVolumePlugin(genericListers configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (ret map[string]interface{}, errs []error) {
+	volumePluginPath := []string{"extendedArguments", "external-cloud-volume-plugin"}
+	defer func() {
+		ret = configobserver.Pruned(ret, volumePluginPath)
+	}()
+	prevObservedConfig := map[string]interface{}{}
+
+	currentCloudVolumePlugin, _, err := unstructured.NestedString(existingConfig, volumePluginPath...)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	if len(currentCloudVolumePlugin) > 0 {
+		if err := unstructured.SetNestedField(prevObservedConfig, currentCloudVolumePlugin, volumePluginPath...); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	listers := genericListers.(configobservation.Listers)
+	infrastructure, err := listers.InfrastructureLister().Get("cluster")
+	if err != nil {
+		return existingConfig, append(errs, err)
+	}
+
+	external, err := isCloudProviderExternal(listers, infrastructure)
+	if err != nil {
+		return existingConfig, append(errs, err)
+	}
+
+	observedConfig := map[string]interface{}{}
+	cloudProvider := getPlatformName(infrastructure.Status.PlatformStatus.Type, recorder)
+
+	// If the cloud provider is external, we should set the option, else leave it empty.
+	if external && len(cloudProvider) > 0 {
+		if err := unstructured.SetNestedField(observedConfig, cloudProvider, volumePluginPath...); err != nil {
+			recorder.Warningf("ObserveCloudVolumePlugin", "Failed setting cloudVolumePlugin: %v", err)
+			return existingConfig, append(errs, err)
+		}
+	}
+
+	if !equality.Semantic.DeepEqual(prevObservedConfig, observedConfig) {
+		recorder.Event("ObserveCloudVolumePlugin", "observed change in config")
+	}
+
+	return observedConfig, errs
+}
+
+// isCloudProviderExternal determines whether or not the external cloud provider should be running within the cluster.
+func isCloudProviderExternal(listers configobservation.Listers, infrastructure *configv1.Infrastructure) (bool, error) {
+	featureGate, err := listers.FeatureGateLister().Get("cluster")
+	if errors.IsNotFound(err) {
+		// When no FG exists, defer to IsCloudProviderExternal to determine if the platform is external
+	} else if err != nil {
+		return false, err
+	}
+
+	external, err := cloudprovider.IsCloudProviderExternal(infrastructure.Status.PlatformStatus, featureGate)
+	if err != nil {
+		return false, err
+	}
+
+	return external, nil
+}
+
+// getPlatformName returns the name of the cloud volume plugin if one is supported for the platform type.
+func getPlatformName(platformType configv1.PlatformType, recorder events.Recorder) string {
```

Comment:
Can't this be shared with library-go's version? Seems reasonable to expose that other one instead of copying

Reply:
I was concerned about exposing those methods as eventually the config observer is going to go away, if we add it to library-go as a public method, we will eventually need to deprecate these right? 

I'll pull up a PR exposing those methods and we can discuss on that if there are concerns about the longevity of the methods

Reply:
Created https://github.com/openshift/library-go/pull/1132, if you think that's appropriate to go ahead with, then I'm happy to update this one too